### DISCUSSION TITLE
Embedding–output decoupling observed in Llama-3 (looking for sanity check / replication)

### DISCUSSION THREAD

**Discussion by buk81** [buk81/uniformity-asymmetry#1] [Show and tell]
Hi everyone,

I’m sharing an exploratory observation from an embedding evaluation experiment and would appreciate a sanity check or replication attempts.

### Context
I developed a metric called **Uniformity Asymmetry (UA)** to measure how tightly LLM embeddings cluster semantically equivalent but differently framed statements.

I tested 4 models (Llama-3.1-8B, Gemma-2-9B, Mistral-7B, Apertus-8B) on **230 paired statements**, and then validated the embedding signal against **actual output log-probabilities**.

### Observation

<img width="1178" height="988" alt="scientific_facts_ua_vs_output" src="https://github.com/user-attachments/assets/9d4d1e62-503a-471c-9cf8-2186fa6f2c9b" />
_Scientific Facts subset: Uniformity Asymmetry (UA, ×10) vs output preference measured as mean ΔNLL between paired framings (1.0 ≈ no preference)._

For **Llama-3.1-8B**, I consistently observe:
- **Near-zero UA** (embedding space looks neutral)
- **Strong output-level preference** across equivalent framings (measured via mean ΔNLL)

Other models (Gemma, Mistral, Apertus) show more consistent embedding–output alignment, though with different patterns.

This suggests a possible **embedding–behavior decoupling** in Llama-3 under this setup.

### What I’m looking for
- Does this experimental setup make sense to you?
- Are there obvious confounds or failure modes I should control for?
- Has anyone observed similar embedding/output mismatches?
- Suggestions for better baselines or null tests are very welcome.

### Resources
- Repo (code, data, figures): https://github.com/buk81/uniformity-asymmetry
- Zenodo DOIs:
  - Paper: 10.5281/zenodo.18110161
  - Extended results: 10.5281/zenodo.18117757

Thanks for any feedback or pointers.

**Comment by kevin-pw** [comment]
I saw your post on reddit. Your experimental setup looks interesting so I look forward to your findings.

To answer your question "Has anyone checked whether Llama-3 embeddings actually predict output behavior": Yes I have and yes they do. The key ingredient: Chat templates!

Since chat models are trained such that text inputs are always formatted using the model-specific chat template, you need to apply those same chat templates to your input text when generating embeddings. Without applying the correct formatting, chat models generate nonsense text and therefore nonsense embeddings.

I see that you are using `transformers` for embedding generation. I have found that the `/embeddings` endpoint provided by vLLM is more convenient for embedding generation with chat models because you can pass a `messages` object with your API request to apply the correct chat template server side.

Good luck!

**Reply by buk81** [reply]
Thanks for this — that’s an excellent point.

You’re absolutely right that chat-tuned models require proper templating to produce in-distribution representations, and this is a concrete confound that should be controlled for.

Clarification on the current setup:
For the cross-model validation I’m running now, I’m intentionally using only base checkpoints (e.g., Meta-Llama-3-8B base, Pythia base) to avoid template-related effects.
I’ll double-check whether the original Llama-3.1-8B runs in the paper used a base or an instruction checkpoint, and rerun with proper chat templating if needed.

Follow-up plan:

- Confirm the checkpoint type used originally.
- If instruct: rerun with apply_chat_template() (or vLLM /embeddings)
- Compare base vs instruct + template vs instruct without template.

Both outcomes are informative:

- If the effect disappears → template-dependent behavior (clear scope clarification)
- If it persists → template-invariant embedding–output decoupling (stronger finding)

Thanks also for the vLLM /embeddings tip — that’s very helpful for the instruct runs.
I’ll report back once the control experiment is complete.

**Reply by buk81** [reply]
Hi @kevin-pw,

just a quick follow-up to say thanks again for flagging the chat-template confound — it turned out to be very important.

We’ve now completed a dedicated control experiment comparing Llama-3.1-8B-Instruct with and without proper chat templating, and the results are written up in a separate follow-up paper (Paper #2):

https://doi.org/10.5281/zenodo.18142454

In short: the late-layer embedding–output inversion persists regardless of templating, but early-stage representations are strongly template-sensitive. With correct chat formatting, the instruct model enters a uniformly negative correlation regime earlier, whereas without templating it behaves more similarly to the base model.

This confirmed exactly the kind of confound you pointed out — the effect doesn’t disappear, but its early-layer behavior changes substantially depending on deployment-format inputs.

Thanks again for the methodological guidance; it significantly strengthened the analysis.
If you have any thoughts on how to best interpret the early-layer template sensitivity, I’d be very interested to hear them.

**Reply by kevin-pw** [reply]
I had a brief look at your paper - the asymmetry inversion in the later layers is interesting and I agree that this finding supports your conclusion around "late-stage decision commitment".

But I am curious if your validation shows that the grounding statement pairs explicitly do not exhibit that same asymmetry. Perhaps you could include the visualization of your validation experiments in your results section. I think showing the neutral symmetry for neutral statement pairs, and then showing the effect of non-neutral statement pairs causing asymmetric geometry would more clearly relate cause and effect.

And I think the early-layer template sensitivity aligns with our understanding that neural networks generally encode smaller-scale structures (such as distinct formatting tokens) in early layers and larger-scale or more complex structures in later layers. Essentially, I think the model can somewhat compensate for the lack of formatting in later layers - but I am basing this interpretation on my experience with observing comparatively simple dense networks, so I am not sure if this statement holds for transformer architectures.

**Reply by buk81** [reply]
Hi @kevin-pw,

Thanks for the thoughtful comments — that's a very good suggestion, and you're right to push on the cause-and-effect question.

We ran a dedicated control comparing **neutral statement pairs** (non-controversial factual pairs) against **political pairs** (opposing stances on the same topic). Interestingly, the result was *not* what we initially expected:

![Neutral vs Political Asymmetry](https://raw.githubusercontent.com/buk81/uniformity-asymmetry/main/Results/neutral_vs_political_asymmetry.png)

Neutral pairs do **not** show lower asymmetry; if anything, they show comparable or slightly higher variance, with no significant difference overall (p = 0.66).

The key reason *seems to be* **semantic distance** rather than topic sensitivity: neutral pairs often compare entirely different facts (far apart in embedding space), whereas political opposites typically discuss the *same concept* with opposite valence — and therefore remain close in representation space.

This helped clarify the interpretation: **Uniformity Asymmetry captures structural/geometric differences in embedding space**, and is most meaningful when applied to *semantically parallel* pairs. We've added this as a caveat.

On your second point about early-layer template sensitivity — I agree with your interpretation. The behavior is consistent with early layers encoding surface-level structure, while later layers focus on more abstract representations.

Really appreciate you pushing on these points — it helped sharpen both the analysis and the framing.

I'd be curious whether this interpretation matches what you've seen in other embedding or probing setups.

Raw data: [Results/neutral_control_test_results.json](https://github.com/buk81/uniformity-asymmetry/blob/main/Results/neutral_control_test_results.json)

Best,
Davide (@buk81)
